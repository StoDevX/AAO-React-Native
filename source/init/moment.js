// @flow
import moment from 'moment-timezone'

// These next values made possible by
// http://stackoverflow.com/questions/42513276

// Tell Moment to round up, so that Moment will always report the
// time until something else in a way that matches the clock time.
// That is, at 5:22pm, Moment will report "8 minutes until 5:30pm"
// until the clock ticks over to 5:23pm.
// If we do Math.floor, we go the other direction: it will report
// "7 minutes until 5:30pm" as soon as 5:22pm starts.
moment.relativeTimeRounding(Math.ceil)

// Tell Moment that a minute is 60 seconds, for reporting purposes.
// It defaults to 45 seconds.
moment.relativeTimeThreshold('s', 60)
