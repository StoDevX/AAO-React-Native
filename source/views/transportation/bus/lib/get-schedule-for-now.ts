import type {BusSchedule, DayOfWeek} from '../types'
import type {Moment} from 'moment-timezone'

const allDaysOfWeek: DayOfWeek[] = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa']

export function getScheduleForNow(
	schedules: BusSchedule[],
	now: Moment,
): BusSchedule {
	// now.day returns 0-6, Sunday to Saturday
	let thisWeekday = allDaysOfWeek[now.day()]

	if (!thisWeekday) {
		throw new Error('Invalid weekday')
	}

	let schedule = schedules.find((instance) =>
		instance.days.includes(thisWeekday),
	)

	if (!schedule) {
		return {
			days: [thisWeekday],
			timetable: [],
			coordinates: {},
			stops: [],
			times: [],
		}
	}

	return schedule
}
