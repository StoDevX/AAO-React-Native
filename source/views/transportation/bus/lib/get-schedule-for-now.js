// @flow
import type {BusSchedule, DayOfWeek} from '../types'
import moment from 'moment-timezone'

const allDaysOfWeek: DayOfWeek[] = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa']

export function getScheduleForNow(
  schedules: BusSchedule[],
  now: moment,
): BusSchedule {
  // now.day returns 0-6, Sunday to Saturday
  const thisWeekday = allDaysOfWeek[now.day()]

  const schedule = schedules.find(schedule =>
    schedule.days.includes(thisWeekday),
  )

  if (!schedule) {
    return {
      days: [thisWeekday],
      timetable: [],
      coordinates: {},
      stops: [],
      times: [],
    }
  }

  return schedule
}
