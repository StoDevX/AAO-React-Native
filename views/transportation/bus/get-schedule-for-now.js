// @flow
import type {BusScheduleType} from './types'
import moment from 'moment-timezone'

const allDaysOfWeek = ['su', 'mo', 'tu', 'we', 'th', 'fr', 'sa']

export default function getScheduleForNow(
  schedules: BusScheduleType[],
  now: typeof moment
): BusScheduleType|void {
  const thisWeekday = now.day() // 0-6, Sunday to Saturday.

  return schedules.find(schedule => {
    return schedule.days
      .map(day => allDaysOfWeek.indexOf(day)) // turn days into numbers
      .includes(thisWeekday)
  })
}
