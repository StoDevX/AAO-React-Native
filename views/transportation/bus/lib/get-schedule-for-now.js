// @flow
import type {BusScheduleType, DayOfWeekType} from '../types'
import moment from 'moment-timezone'

const allDaysOfWeek: DayOfWeekType[] = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa']

export function getScheduleForNow(
  schedules: BusScheduleType[],
  now: typeof moment
): BusScheduleType|void {
  const thisWeekday = now.day() // 0-6, Sunday to Saturday.

  return schedules.find(schedule => {
    return schedule.days
      .map(day => allDaysOfWeek.indexOf(day)) // turn days into numbers
      .includes(thisWeekday)
  })
}
