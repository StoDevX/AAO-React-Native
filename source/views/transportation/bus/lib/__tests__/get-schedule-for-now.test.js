// @flow
import {getScheduleForNow} from '../get-schedule-for-now'
import moment from 'moment'

import type {BusScheduleType} from '../../types'
function buildBusSchedules(): BusScheduleType[] {
  return [
    {
      days: ['Mo', 'Tu'],
      stops: ['St. Olaf', 'Carleton', 'Food Co-op', 'Cub/Target', 'El Tequila', 'Food Co-op', 'Carleton', 'St. Olaf'],
      times: [['4:15pm',  '4:22pm',   '4:23pm',     '4:33pm',     '4:37pm',     '4:43pm',     '4:44pm',   '4:52pm'],
              ['4:55pm',  '5:02pm',   '5:03pm',     '5:13pm',     '5:17pm',     '5:23pm',     '5:24pm',   '5:32pm'],
              ['5:35pm',  '5:42pm',   '5:43pm',     '5:53pm',     '5:57pm',     '6:03pm',     '6:04pm',   '6:12pm'],
      ],
    },
    {
      days: ['We', 'Th'],
      stops: ['St. Olaf', 'Carleton', 'Food Co-op', 'Cub/Target', 'El Tequila', 'Food Co-op', 'Carleton', 'St. Olaf'],
      times: [['6:15pm',  '6:22pm',   '6:23pm',     '6:33pm',     '6:37pm',     '6:43pm',     '6:44pm',   '6:52pm'],
              ['6:55pm',  '7:02pm',   '7:03pm',     '7:13pm',     '7:17pm',     '7:23pm',     '7:24pm',   '7:32pm'],
              ['7:35pm',  '7:42pm',   '7:43pm',     '7:53pm',     '7:57pm',     '8:03pm',     '8:04pm',   '8:12pm'],
      ],
    },
    {
      days: ['Fr', 'Sa'],
      stops: ['St. Olaf', 'Carleton', 'Food Co-op', 'Cub/Target', 'El Tequila', 'Food Co-op', 'Carleton', 'St. Olaf'],
      times: [['8:15pm',  '8:22pm',   '8:23pm',     '8:33pm',     '8:37pm',     '8:43pm',     '8:44pm',   '8:52pm'],
              ['8:55pm',  '9:02pm',   '9:03pm',     '9:13pm',     '9:17pm',     '9:23pm',     '9:24pm',   '9:32pm'],
              ['9:35pm',  '9:42pm',   '9:43pm',     '9:53pm',     '9:57pm',     '10:03pm',    '10:04pm',  '10:12pm'],
      ],
    },
  ]
}


it('returns the bus schedule for today', () => {
  let now = moment('Fri 10:01', 'dddd H:mm')
  let input = buildBusSchedules()
  let actual = getScheduleForNow(input, now)
  if (!actual) {
    return
  }

  expect(actual.days).toEqual(['Fr', 'Sa'])
})

it('returns `undefined` if there is no schedule for today', () => {
  let now = moment('Sun 10:01', 'dddd H:mm')
  let input = buildBusSchedules()
  let actual = getScheduleForNow(input, now)

  expect(actual).toBeUndefined()
})
