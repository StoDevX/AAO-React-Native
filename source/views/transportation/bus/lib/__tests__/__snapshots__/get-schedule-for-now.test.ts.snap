// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns an empty schedule if there is no schedule for today 1`] = `
{
  "coordinates": {},
  "days": [
    "Su",
  ],
  "stops": [],
  "times": [],
  "timetable": [],
}
`;

exports[`returns the bus schedule for today 1`] = `
{
  "coordinates": {},
  "days": [
    "Fr",
    "Sa",
  ],
  "stops": [
    "St. Olaf",
    "Carleton",
    "Food Co-op",
    "Cub/Target",
    "El Tequila",
    "Food Co-op",
    "Carleton",
    "St. Olaf",
  ],
  "times": [
    [
      "2017-11-19T02:15:00.000Z",
      "2017-11-19T02:22:00.000Z",
      "2017-11-19T02:23:00.000Z",
      "2017-11-19T02:33:00.000Z",
      "2017-11-19T02:37:00.000Z",
      "2017-11-19T02:43:00.000Z",
      "2017-11-19T02:44:00.000Z",
      "2017-11-19T02:52:00.000Z",
    ],
    [
      "2017-11-19T02:55:00.000Z",
      "2017-11-19T03:02:00.000Z",
      "2017-11-19T03:03:00.000Z",
      "2017-11-19T03:13:00.000Z",
      "2017-11-19T03:17:00.000Z",
      "2017-11-19T03:23:00.000Z",
      "2017-11-19T03:24:00.000Z",
      "2017-11-19T03:32:00.000Z",
    ],
    [
      "2017-11-19T03:35:00.000Z",
      "2017-11-19T03:42:00.000Z",
      "2017-11-19T03:43:00.000Z",
      "2017-11-19T03:53:00.000Z",
      "2017-11-19T03:57:00.000Z",
      "2017-11-19T04:03:00.000Z",
      "2017-11-19T04:04:00.000Z",
      "2017-11-19T04:12:00.000Z",
    ],
  ],
  "timetable": [
    {
      "departures": [
        "2017-11-19T02:15:00.000Z",
        "2017-11-19T02:55:00.000Z",
        "2017-11-19T03:35:00.000Z",
      ],
      "name": "St. Olaf",
    },
    {
      "departures": [
        "2017-11-19T02:22:00.000Z",
        "2017-11-19T03:02:00.000Z",
        "2017-11-19T03:42:00.000Z",
      ],
      "name": "Carleton",
    },
    {
      "departures": [
        "2017-11-19T02:23:00.000Z",
        "2017-11-19T03:03:00.000Z",
        "2017-11-19T03:43:00.000Z",
      ],
      "name": "Food Co-op",
    },
    {
      "departures": [
        "2017-11-19T02:33:00.000Z",
        "2017-11-19T03:13:00.000Z",
        "2017-11-19T03:53:00.000Z",
      ],
      "name": "Cub/Target",
    },
    {
      "departures": [
        "2017-11-19T02:37:00.000Z",
        "2017-11-19T03:17:00.000Z",
        "2017-11-19T03:57:00.000Z",
      ],
      "name": "El Tequila",
    },
    {
      "departures": [
        "2017-11-19T02:43:00.000Z",
        "2017-11-19T03:23:00.000Z",
        "2017-11-19T04:03:00.000Z",
      ],
      "name": "Food Co-op",
    },
    {
      "departures": [
        "2017-11-19T02:44:00.000Z",
        "2017-11-19T03:24:00.000Z",
        "2017-11-19T04:04:00.000Z",
      ],
      "name": "Carleton",
    },
    {
      "departures": [
        "2017-11-19T02:52:00.000Z",
        "2017-11-19T03:32:00.000Z",
        "2017-11-19T04:12:00.000Z",
      ],
      "name": "St. Olaf",
    },
  ],
}
`;
